language/compiler/ir-to-bytecode/syntax/src/lexer.rs

```diff
@@ -311,41 +132,248 @@ impl<'input, 'builder> Lexer<'input, 'builder> {
         let text = self.text.trim_start();
         let whitespace = self.text.len() - text.len();
         let start_offset = self.consumed + whitespace;
-        if text.is_empty() {
-            self.text = text;
-            self.consumed = start_offset;
-            self.token = (start_offset, Token(Tok::EOF, ""), start_offset);
-            Ok(())
-        } else {
-            let mut longest_match = 0;
-            let mut index: Option<usize> = None;
-            for (i, token_exp) in self.regex_vec.iter().enumerate() {
-                if let Some(m) = token_exp.find(text) {
-                    let len = m.end();
-                    if len >= longest_match {
-                        longest_match = len;
-                        index = Some(i.try_into().unwrap());
+        let c: char = match text.chars().next() {
+            Some(next_char) => next_char,
+            None => {
+                self.text = text;
+                self.consumed = start_offset;
+                self.token = (start_offset, Token(Tok::EOF, ""), start_offset);
+                return Ok(());
+            }
+        };
+        let (tok, len) = match c {
+            '0'..='9' => {
```

Comment:
Maybe some comments above the match for what tokens are covered? Not necessary but might make these ranges easier?

Reply:
I just tried that to see what it would look like. It's marginally useful for this first match but not so much for the rest of them. It is either obvious which tokens are involved or there are too many possible tokens to list them. Maybe I'm misunderstanding your suggestion, but I don't think this would be helpful.